src/map/battle_skill_factory.hpp

```diff
@@ -0,0 +1,98 @@
+#ifndef BATTLE_SKILL_FACTORY_HPP
+#define BATTLE_SKILL_FACTORY_HPP
+
+#include <memory>
+#include <unordered_map>
+#include <functional>
+#include <mutex>
+#include <vector>
+#include <common/cbasetypes.hpp>
+#include "battle_skill.hpp"
+
+class SkillAlreadyImplementedException : public std::logic_error {
+public:
+    explicit SkillAlreadyImplementedException(const std::string &what_arg) : std::logic_error(what_arg) {};
+    explicit SkillAlreadyImplementedException(uint16_t skill_id) : std::logic_error("Skill " + std::to_string(skill_id) + " is already implemented") {};
+};
+
+/**
+ * Factory class for creating and managing BattleSkill instances
+ * Uses singleton pattern with thread-safe skill registration and caching
+ */
+class BattleSkillFactory
+{
+private:
+    std::unordered_map<uint16, std::shared_ptr<BattleSkill>> skill_db;
+    BattleSkillFactory();
+
+public:
+    // Singleton access
+    static BattleSkillFactory &instance();
+
+    // Delete copy constructor and assignment operator
+    BattleSkillFactory(const BattleSkillFactory &) = delete;
+    BattleSkillFactory &operator=(const BattleSkillFactory &) = delete;
+
+    // Destructor
+    ~BattleSkillFactory() = default;
+
+    /**
+     * Register a skill type with template for type safety
+     * @tparam SkillType The skill class type
+     * @param skill_id The skill ID to register
+     */
+    template <typename SkillType>
+    void register_skill(uint16 skill_id)
```

Comment:
Can't we just take a shared_ptr here?
```
void register_skill(uint16 skill_id, const std::shared_ptr<BattleSkill>& skill) {
  skill_db[skill_id] = skill;
}
```
this way, we can save the compiler work having to generate this function for EVERY skill

Reply:
I've added at [f59db4e](https://github.com/rathena/rathena/pull/9468/commits/f59db4e6ce60d13dc7eabea5f60ed0ffce52e8a1)